### ISSUE TITLE
It seems that the ability to identify files is not strong enough

### ISSUE THREAD

**Issue by LLH-l** [ZerBea/hcxtools#227]
I found 43 complete handshake files, but the hcxpcapngtool tool does not seem to recognize them
In order to verify this test, I used other tools to recover one of the correct passwords (which has been verified to be connected to the AP)
I detect message 3 of 4 using Wireshark
You may need to recheck them

![1](https://user-images.githubusercontent.com/80609791/201687586-8a54e70c-406c-42e4-9db5-166af8332f64.png)
[Wireshark_message3of4.zip](https://github.com/ZerBea/hcxtools/files/10003990/Wireshark_message3of4.zip)

**Comment by ZerBea** [comment]
A single EAPOL MESSAGE is not enough to calculate a MESSAGE PAIR.
We need at least 2 EAPOL MESSAGEs (one from the AP and a matching one with correct REPLAY COUNT from the CLIENT).
Absolutely none of the attached dump files satisfy this requirements:

example one: 1_(2456).cap
```
$ tshark -r '1_(2456).cap' -Y "eapol" -T fields -e frame.number -e frame.time -e wlan_rsna_eapol.keydes.msgnr -e eapol.keydes.replay_counter
963		Aug 30, 2019 15:48:28.335373000 CEST	2	1
966		Aug 30, 2019 15:48:28.342538000 CEST	4	2
2825	Aug 30, 2019 15:49:05.172038000 CEST	3	2
```
EAPOL messages not in a row 2, 4, 2)

example two:
```
$ hcxpcapngtool 1_(273).cap
hcxpcapngtool 6.2.7-73-gedd0151 reading from 1_(273).cap...

summary capture file
--------------------
file name................................: 1_(273).cap
version (pcap/cap).......................: 2.4 (very basic format without any additional information)
timestamp minimum (GMT)..................: 14.08.2019 11:07:43
timestamp maximum (GMT)..................: 14.08.2019 11:08:33
used capture interfaces..................: 1
link layer header type...................: DLT_IEEE802_11_RADIO (127)
endianess (capture system)...............: little endian
packets inside...........................: 252
frames with correct FCS..................: 237
packets received on 2.4 GHz..............: 252
EAPOL messages (total)...................: 249
EAPOL RSN messages.......................: 250
EAPOLTIME gap (measured maximum usec)....: 6315
EAPOL ANONCE error corrections (NC)......: not detected
REPLAYCOUNT gap (measured maximum).......: 206
EAPOL M1 messages (total)................: 244
EAPOL M2 messages (total)................: 1
EAPOL M3 messages (total)................: 2
EAPOL M4 messages (total)................: 2
EAPOL pairs (total)......................: 2
malformed packets (total)................: 2
EAPOL messages (malformed packets).......: 8
```
This dump file contaain a complete 4way handshake
249	Aug 14, 2019 11:08:33.960877000 CEST	1	1
250	Aug 14, 2019 11:08:33.967192000 CEST	2	1
251	Aug 14, 2019 11:08:33.970330000 CEST	3	2
252	Aug 14, 2019 11:08:33.973729000 CEST	4	2

But the ESSID is missing in the dumpfile. We need the ESSID to calculate PBKDF2.

example three:
```
$ tshark -r '1_(5997).cap' -Y "eapol" -T fields -e frame.number -e frame.time -e wlan_rsna_eapol.keydes.msgnr -e eapol.keydes.replay_counter
1734	Feb 17, 2022 15:18:03.659462000 CET	3	2
16951	Feb 17, 2022 15:18:36.524300000 CET	1	1
16954	Feb 17, 2022 15:18:36.528394000 CET	3	2
29295	Feb 17, 2022 15:19:10.325624000 CET	2	1
29297	Feb 17, 2022 15:19:10.329722000 CET	4	2
```
EAPOL messages not in a row 3, 1, 3, 2, 4


example 4:
```
$ tshark -r '2_(188).cap' -Y "eapol" -T fields -e frame.number -e frame.time -e wlan_rsna_eapol.keydes.msgnr -e eapol.keydes.replay_counter
1163	Oct  7, 2022 08:05:19.941584000 CEST	1	1
7739	Oct  7, 2022 08:07:40.770065000 CEST	1	1
7744	Oct  7, 2022 08:07:40.797198000 CEST	3	2
12120	Oct  7, 2022 08:08:47.405007000 CEST	1	1
12121	Oct  7, 2022 08:08:47.432145000 CEST	3	2
12122	Oct  7, 2022 08:08:47.434701000 CEST	4	2
17731	Oct  7, 2022 08:10:34.987661000 CEST	2	1
29258	Oct  7, 2022 08:13:20.331790000 CEST	3	2
```
EAPOL MESSAGEs 17731 and 29258 are in a row (2, 3), the REPLAY COUNTER matches (1, 2) but the EAPOL time between them is > thew 3 minutes. They do not belong to the same authentication sequence.

In other words all of this cap files are crappy!

But I want give up.
So please add exactly the cap file of which do you think it meet the conditions to recover the PSK!
The conditions are:
ESSID inside the dump file
M1, M2, M3, M4 in a row inside (no other EAPOL messages between them)
REPLAY COUNTER must match
time between M1 and M4 < 5000ms

**Comment by ZerBea** [comment]
The 4way handshake is explained here:
https://medium.com/wifi-testing/four-way-handshake-32356fbec1b5
This condition must be met to convert the hash (mandatory)!!!!!!!!

Using this parameters
on hcxpcapngtool
```
-all                              : convert all possible hashes instead of only the best one
                                     that can lead to much overhead hashes
                                     use hcxhashtool to filter hashes
                                     need hashcat --nonce-error-corrections >= 8
--eapoltimeout=<digit>             : set EAPOL TIMEOUT (milliseconds)
                                   : default: 5000 ms
--nonce-error-corrections=<digit>  : set nonce error correction
                                     warning: values > 0 can lead to uncrackable handshakes
                                   : default: 0
--ignore-ie                        : do not use CIPHER and AKM information
                                     this will convert all frames regadless of
                                     CIPHER and/OR AKM information,
                                     and can lead to uncrackable hashes
```

in combination with hashcat
`     --nonce-error-corrections  | Num  | The BF size range to replace AP's nonce last bytes   | --nonce-error-corrections=16
`
you can lower this restrictions, but that may waste your GPU time due to false MESSAGE PAIRs.

**Comment by ZerBea** [comment]
And last, but not least: hcxpcapngtool told you that the dump files are crapy!
```
Information: limited dump file format detected!
This file format is a very basic format to save captured network data.
It is recommended to use PCAP Next Generation dump file format (or pcapng for short) instead.
The PCAP Next Generation dump file format is an attempt to overcome the limitations
of the currently widely used (but very limited) libpcap (cap, pcap) format.
https://www.wireshark.org/docs/wsug_html_chunked/AppFiles.html#ChAppFilesCaptureFilesSection
https://github.com/pcapng/pcapng

Information: radiotap header is missing!
Radiotap is a de facto standard for 802.11 frame injection and reception.
The radiotap header format is a mechanism to supply additional information about frames,
from the driver to userspace applications.
https://www.radiotap.org/

Warning: out of sequence timestamps!
This dump file contains frames with out of sequence timestamps.
That is a bug of the capturing tool.

Warning: too many deauthentication/disassociation frames detected!
That can cause that an ACCESS POINT change channel, reset EAPOL TIMER,
renew ANONCE and set PMKID to zero.
This could prevent to calculate a valid EAPOL MESSAGE PAIR
or to get a valid PMKID.



Information: missing frames!
This dump file does not contain undirected proberequest frames.
An undirected proberequest may contain information about the PSK.
It always happens if the capture file was cleaned or
it could happen if filter options are used during capturing.
That makes it hard to recover the PSK.
```

That means that it is not possible to convert them by default options!
In every thread I wrote that hcxtools are analyses tools which are designed to be used by experienced users.
This means you need to know what you are doing in this case (options to lower the restrictions), e.g.:

```
$ hcxpcapngtool --all --nonce-error-corrections=1024 --eapoltimeout=100000000 -o test.22000 *.cap
$ hashcat -m 22000 --nonce-error-corrections=1024 test.22000 Password.tx
hashcat (v6.2.6-111-g3e497ac87) starting
...
4e896f25666821ffc8887a9ac0f47aa5:a87d129981ac:5870c6040367:zl:jj..1234
                                                          
Session..........: hashcat
Status...........: Exhausted
Hash.Mode........: 22000 (WPA-PBKDF2-PMKID+EAPOL)
Hash.Target......: test.22000
Time.Started.....: Mon Nov 14 18:55:45 2022 (6 secs)
Time.Estimated...: Mon Nov 14 18:55:51 2022 (0 secs)
Kernel.Feature...: Pure Kernel
Guess.Base.......: File (Password.txt)
Guess.Queue......: 1/1 (100.00%)
Speed.#1.........:        6 H/s (1.41ms) @ Accel:4 Loops:256 Thr:512 Vec:1
Recovered........: 1/103 (0.97%) Digests (total), 1/103 (0.97%) Digests (new), 1/35 (2.86%) Salts
Progress.........: 35/35 (100.00%)
Rejected.........: 0/35 (0.00%)
Restore.Point....: 1/1 (100.00%)
Restore.Sub.#1...: Salt:34 Amplifier:0-1 Iteration:0-1
Candidate.Engine.: Device Generator
Candidates.#1....: jj..1234 -> jj..1234
Hardware.Mon.#1..: Temp: 48c Fan: 33% Util: 88% Core:1898MHz Mem:5005MHz Bus:16

Started: Mon Nov 14 18:55:44 2022
Stopped: Mon Nov 14 18:55:52 2022
```

Closed this again, because the MESSAGE PAIR is converted by hcxpcapngtool and the PSK is recovered by hashcat.
4e896f25666821ffc8887a9ac0f47aa5:a87d129981ac:5870c6040367:zl:jj..1234

**Comment by ZerBea** [comment]
BTW:
hcxpmktool confirm this, too:
```
$ hcxpmktool -l WPA*02*4e896f25666821ffc8887a9ac0f47aa5*a87d129981ac*5870c6040367*7a6c*210e79ce58cf3b22702b507595637e325b26f2b8f5297bae14c453bb704e0517*0103007502010a00000000000000000000e9ad90d39b50c64c3efe4dfe66441336afde41b9fa028e556865ab1eb23efefe000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001630140100000fac040100000fac040100000fac020000*82 -p "jj..1234"

HASH FORMAT.: EAPOL (WPA*02)
ESSID.......: zl
MAC_AP......: a87d129981ac
MAC_CLIENT..: 5870c6040367
PSK.........: jj..1234
PMK.........: 2648563550eb5a020ebd7aaf22cd7d0a0d6fc342716db322a273bf3a875aa78d (calculated)
KEY VERSION.: WPA2
NONCE AP....: 210e79ce58cf3b22702b507595637e325b26f2b8f5297bae14c453bb704e0517
NONCE CLIENT: e9ad90d39b50c64c3efe4dfe66441336afde41b9fa028e556865ab1eb23efefe
KCK.........: 1d781e9fc0a35e7eaaefc0e9d9d87106 (calculated)
KEK.........: a84fada36f6e005870c6040367a87d12 (calculated)
TK..........: 9981ac210e79ce58cf3b22702b507595 (calculated)
TKIP TX MIC.: 637e325b26f2b8f5 (calculated)
TKIP RX MIC.: 297bae14c453bb70 (calculated)
MIC.........: 4e896f25666821ffc8887a9ac0f47aa5 (confirmed)
PMKID.......: 2f2b9c27ea244d5d04ca106f438fc39e (calculated)
```
The Plain Master Key and all session keys are calculated as expected. That include the PMKID, too.

**Comment by LLH-l** [comment]
Because the biological clock in my region is different, I will soon go to sleep and can't reply to you in time,
But I always think they are valuable. Although the file quality is poor, there is at least one correct password in it, so I try to recover the password,
The successful recovery of the correct password can explain a problem. This experiment can at least prove some defects

**Comment by ZerBea** [comment]
Please notice that hcxpcapngtool default options are designed to get the best MESSAGE PAIR from a dump file.
If you don't think it is the best one, you can increase the restrictions by using this options.

Conversely: If no MESSAGE PAIR is converted due to poor quality of the dump files by default options, you can lower the restrictions by the same options as I did it.
In that case it is mandatory that you know that this will produce a lot of invalid MESSAGE PAIRs, too and running hashcat on them will waste much GPU time.
Only one MESSAGE PAIR is valid - 102 are possible garbage, because they are converted from garbage dump files 
Recovered........: 1/103 (0.97%) Digests (total), 1/103 (0.97%) Digests (new), 1/35 (2.86%) Salts
This is the worst case, but it allow to handle crappy dump files.

This is the best case:
Recovered........: 1/1 (100.00%) Digests (total), 1/1 (100.00%) Digests (new)
Only one hash of high quality and you can be sure that you don't waste GPU time. 

Lowering the default restrictions produce garbage and that is absolutely a no go for online hash recovering like:
https://wpa-sec.stanev.org
or online conversion sites as
https://hashcat.net/cap2hashcat/

Please take a closer look at hashcat online converter. It is mentioned there, too:

`The online converter works exclusively with default settings. Any additional in-depth tuning exceeds the scope of this online service.
`
and
```
For best results, avoid tools that strip or modify capture files, such as:

    airodump-ng (with filter options)
    besside-ng
    wpaclean
    old bettercap versions
    old pwnagotchi versions
    tshark (with filter options)
    wireshark (with filter options)
```

**Comment by LLH-l** [comment]
OK, thanks

**Comment by ZerBea** [comment]
No problem, you're welcome.
The entire WiFi stuff is really hard to understand. That starts with 802.11 protocol specifications, end up with hash calculation and include radio technology (channel spectrum and coding).
But once you understood it, you can do magic.
